### PR TITLE
Initial implementation for tsconfig-generator

### PR DESCRIPTION
**Pull request by cdagli**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,19 @@
...
+```
+npm install tsconfig-generator
+```
+
+```
+Usage: index [options]
+
+  Options:
+
+    -V, --version              output the version number
```
Comment (bgedik) [inline README.md]: Any reason for usıng capital V?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,19 @@
 # tsconfig-generator
-Creates a tsconfig.json by using the paths defined with environment variables
+Creates a tsconfig.json by using the paths defined with environment variables. Use the pattern `${ENV_VARIABLE}` for
```
Comment (bgedik) [inline README.md]: Creates a tsconfig.json file from a template configuration that includes references to environment variables. Use the `${ENV_VARIABLE}` pattern for substituting environment variable values as a string, or the  `#{ENV_VARIABLE}` pattern for substituting environment variable values as a relative path.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+const DEFAULT_TS_CONFIG_FILE_PATH = './tsconfig.json';
+const DEFAULT_JSON_INDENTATION = 4;
+const environmentVariableRegex = new RegExp(/\$\{([\w]+)\}/, 'g');
+const pathVariableRegex = new RegExp(/#\{([\w]+)\}/, 'g');
+
+program
+    .version('0.0.2')
+    .option('-i, --input [inputFile]', 'Input JSON file path')
+    .option('-o, --output [outputFile]', 'Output JSON file path (Default: ./tsconfig.json)')
```
Comment (bgedik) [inline index.js]: specifies the path of the output JSON file

File: index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            if (typeof config[key] === 'string')
+            {
+                config[key] = replaceMatchedAll(environmentVariableRegex, config[key], outputFilePath);
+                config[key] = replaceMatchedAll(pathVariableRegex, config[key], outputFilePath);
+            }
+        }
+    }
+}
+
+function replaceMatchedAll(regex, text, outputFilePath)
```
Comment (bgedik) [inline index.js]: ``replaceAllMatches``?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+{
+    let matched = null;
+
+    do {
```
Comment (bgedik) [inline index.js]: there is an easier way, I think:
```
text.replace(regex, function(substring, variableName)
{
     // just return the replacement
});

File: index.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        matched = regex.exec(text);
+        if (matched)
+        {
+            const environmentVariableValue = process.env[matched[1]];
+            if (!environmentVariableValue)
+            {
+                console.log(`WARNING: Environment variable '${matched[1]}' has no value. Skipping.`);
```
Comment (bgedik) [inline index.js]: let's make this an error

File: index.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            }
+            else if (regex === pathVariableRegex)
+            {
+                text = text.replace(matched[0], path.relative(path.dirname(outputFilePath),
+                    environmentVariableValue));
+            }
+            else if (regex === environmentVariableRegex)
+            {
+                console.log(
+                    `WARNING: '${matched[1]}' is not a valid path. Using the actual value '${environmentVariableValue}'.`
```
Comment (bgedik) [inline index.js]: let's make this an error

File: README.md (hunk 2 above)
Comment (cdagli) [reply README.md]: Done

File: index.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+const program = require('commander');
+
+const DEFAULT_TS_CONFIG_FILE_PATH = './tsconfig.json';
+const DEFAULT_JSON_INDENTATION = 4;
+const environmentVariableRegex = new RegExp(/\$\{([\w]+)\}/, 'g');
+const pathVariableRegex = new RegExp(/#\{([\w]+)\}/, 'g');
+
+program
+    .version('0.0.2')
+    .option('-i, --input [inputFile]', 'Input JSON file path')
```
[the remark this answers is not in the document]
Comment (cdagli) [reply index.js]: Done

File: index.js (hunk 4 above)
Comment (cdagli) [reply index.js]: Done

File: index.js (hunk 5 above)
Comment (cdagli) [reply index.js]: Much better. Fixed.

File: index.js (hunk 6 above)
Comment (cdagli) [reply index.js]: Done. Also added a test.

File: index.js (hunk 7 above)
Comment (cdagli) [reply index.js]: Removed this one.

**Pull request merged**